Add instrumentation using XRay

pamarcos opened the pull request:
<!--- Disable AI PR formatting assistant: true -->
### Changelog category (leave one):

- New Feature

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):

Add instrumentation at runtime using XRay to debug issues in production and profile deterministically. Resolves #74249

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

### Details

Add instrumentation to ClickHouse using LLVM's XRay feature to debug issues in production without needing to rebuild. This is a reworked PR to bring the work that @htuall started at https://github.com/ClickHouse/ClickHouse/pull/77153 to be production-ready.

Please take a look at the [original issue](#74249) to understand better what this is about. The TL;DR is that we can leverage LLVM's XRay to add a prolog and an epilog when entering and exiting functions that are longer than 200 instructions (by default) to execute extra handlers.

- Enable the XRay instrumentation for Linux x86_64 and aarch64 for all non-sanitized builds. Using XRay along with other sanitizers is incompatible. See [this comment](https://github.com/ClickHouse/ClickHouse/pull/77153#issuecomment-3425709837).
- New `SYSTEM INSTRUMENT` statement support to add and remove instrumentation points
- New system table to inspect instrumented symbols: `system.instrumentation`
- Add new columns to `system.trace_log`  to query instrumentation log entries
- Read the `xray_instr_map` only when there's a need to patch a function
- 3 types of handlers so far: LOG, SLEEP and PROFILE. For the first two, you can install them on `entry` or on `exit` of the function

Closes https://github.com/ClickHouse/ClickHouse/issues/74249

Binary size comparison shows a ~4.5% increase for x86_64 and ~7% increase for aarch64.

htuall committed:
working on parsing xray_instr_map section from elf-binary to map xray funcIds to function names

htuall committed:
new logic for mapping functionIDs to their names + new experiment with custom handler for xray + started working on adding SYSTEM-statements to add/remove instrumentation(wip)

htuall committed:
created XRayInstrumentationManager class + finished work on system statements using it + attempted to add system.instrumentation table(not sure if it's coorrect

htuall committed:
fix style + fix build

htuall committed:
fix build

htuall committed:
fix style

htuall committed:
attempt to fix build: does not find llvm files

htuall committed:
revert

htuall committed:
try to fix llvm includes

pamarcos committed:
Merge branch 'master' into xray_experiments

pamarcos committed:
Only build XRay-related instrumentation when XRay is enabled

Let's build with ENABLE_XRAY=1 only for the Debug build on AMD
for now. This is just to ensure on the CI that it builds
correctly.

There are some USE_XRAY missing in other files, but it is left
on purpose as an excercise to the reader :)

pamarcos committed:
Fix XRay includes

Commit:
only instrument if USE_XRAY=1

htuall committed:
fix unterminated conditional directive error + replaced ifdef with if(needs check) + includes to cpp from header

htuall committed:
fix return type

htuall committed:
add one more if USE_XRAY

htuall committed:
add include in cpp + if USE_XRAY

htuall committed:
add new system statements to show privileges test

htuall committed:
Merge branch 'master' into xray_experiments

htuall committed:
fix includes in XRayInstrumentationManager.h

htuall committed:
Merge branch 'xray_experiments' of github.com:htuall/ClickHouse into xray_experiments

htuall committed:
use correct xray function

htuall committed:
fix System/SystemStorageInstrumentation + fix style

htuall committed:
started working on system.instrumentation_profiling_log, in progress

pamarcos committed:
Fix XRay linking error

htuall committed:
define getColumnsDescription

htuall committed:
fix style

htuall committed:
fix new system statement: more intuitive order of parameters, parse handler_name and function_name as identifiers and not string literals, add possibility to pass different type of parameters for differnt hadlers and make them optional

htuall committed:
fix build errors

htuall committed:
fix parameters parsing

htuall committed:
working on InstrumentationProfilingLog

htuall committed:
added logic for stripping function names + throw exception for invalid handler or function + make parameters optional + do not need handler and parameters for remove

htuall committed:
fixed parsing but need to consider a better way to do this, rn seems like a duct tape

htuall committed:
never instrument instrumentWithXRay function

htuall committed:
added new handlers + pass and validate  parameters

htuall committed:
fix build

htuall committed:
fix style

htuall committed:
write to system.instrumentation_profiling_log in profile handler

htuall committed:
make system.instrumentation_profilling_log convertable to Chrome trace format

htuall committed:
fix style

htuall committed:
add shutdown

pamarcos committed:
Revert "add shutdown"

This reverts commit f887e27bf80ffe5d2b40e51c0e8a2291fafc99e4.

pamarcos committed:
Tune fields needed for system.instrumentation_profiling_log

pamarcos committed:
Clarify column name

htuall committed:
dispatch handler + new columns in system.instrumentation

htuall committed:
fix style

htuall committed:
fix build

htuall committed:
fix linker error

htuall committed:
small fixes

htuall committed:
small fix for profile

htuall committed:
small fix for system.instrumentation_profiling_log

htuall committed:
clean up code

htuall committed:
fix style

pamarcos committed:
Merge branch 'master' into xray_experiments

pamarcos committed:
Add XRay prefix to classes and tables

At first I thought not having the XRay was better, but I've
changed my mind. I prefer to make it very explicit. Also,
this way it's easier to filter what's related to XRay
instrumentation and what's not.

htuall committed:
small changes about style + stop using call once + correctly set dispatch handler now

htuall committed:
updates

htuall committed:
Merge branch 'master' into xray_experiments

htuall committed:
accept handler names both in lower and upper case, use scope_exit for in_hook variable, use TSA_GUARDED_BY, fixed mutex usage -- less mutex locks and only when needed + one mutex for all three fields

htuall committed:
Merge branch 'master' into xray_experiments

htuall committed:
unit tests for removeTemplateArgs and extractNearestNamespaceAndFunction

htuall committed:
Merge branch 'xray_experiments' of github.com:htuall/ClickHouse into xray_experiments

pamarcos committed:
Use ClickHouse's SCOPE_EXIT instead of LLVM's one

pamarcos committed:
Avoid using static members when they're not needed

pamarcos committed:
Fix errors raised by TSA

pamarcos committed:
Merge branch 'master' into xray_experiments

pamarcos committed:
Bring back getInstrumentedFunctions

pamarcos committed:
Fix linter for unit test

pamarcos committed:
Fix removeTemplateArgs

pamarcos committed:
Fix extractNearestNamespaceAndFunction

pamarcos committed:
Fix using uppercase handler name

pamarcos committed:
Improve log

pamarcos committed:
Fix system instrument access

pamarcos committed:
Another fix for lowercase instrument

pamarcos committed:
Improve SYSTEM INSTRUMENT REMOVE to use instrumentation_point_id

This is much simpler, since we can use the unique ID for each
instrumentation point, instead of having to specify the function
name and the handler.

pamarcos committed:
Attempt to fix clang-tidy error in test

pamarcos committed:
Fix for real the clang-tidy error in unit test

pamarcos committed:
Rename instrumentation_point_id to just id

pamarcos committed:
Add INSTRUMENT REMOVE ALL option

pamarcos committed:
Add SYSTEM INSTRUMENT ADD/REMOVE test

pamarcos committed:
Merge branch 'master' into xray_experiments

pamarcos committed:
Add SYSTEM INSTRUMENT LOG and SLEEP test

pamarcos committed:
Fix style check

pamarcos committed:
Restore original name cause it's causing issues in sync

pamarcos committed:
Make the test faster

pamarcos committed:
Small improvements. Hold the mutex the minimum

pamarcos committed:
Add SYSTEM INSTRUMENT PROFILE test

pamarcos committed:
Consolidate the logger

pamarcos committed:
Consolidate 3 different maps with boost::multi_index

pamarcos committed:
Parse the XRay instrumentation map in parallel jobs

pamarcos committed:
Fix style check

pamarcos committed:
Enable XRay for amd_release and amd_binary

pamarcos committed:
Fix stripped function names not being unique

pamarcos committed:
Fix job size

pamarcos committed:
Add stack trace to LOG handler

pamarcos committed:
Add entry/exit option for LOG and SLEEP handlers

- Big refactor to avoid having different data structures that hold
  similar maps for instrumented points.

pamarcos committed:
Fix tests and issues found after refactor

pamarcos committed:
Merge branch 'master' into xray_experiments

pamarcos committed:
Fix USE_XRAY guard and linter

pamarcos committed:
Fix build for non-XRay flavor

pamarcos committed:
Add test for ENTRY/EXIT

pamarcos committed:
Make handler order deterministic

pamarcos committed:
Fix config with xray_instrumentation_profiling_log table

pamarcos committed:
Merge branch 'master' into xray_experiments

pamarcos committed:
Fix XRay instrumentation for tail exit

pamarcos committed:
Merge branch 'master' into xray_experiments

pamarcos committed:
Improve text in case function is not found

pamarcos committed:
Profile before sleep handler

pamarcos committed:
Enable XRay for all Linux flavors

pamarcos committed:
Merge branch 'master' into xray_experiments

pamarcos committed:
Remove XRay for sanitizers because clang_rt symbols clash

pamarcos committed:
Merge branch 'master' into xray_experiments

pamarcos committed:
Add documentation for the XRay's instrumentation feature

pamarcos committed:
Use Int32 everywhere for function_id to be consistent with XRay's signature

pamarcos committed:
Add words to dictionary

pamarcos committed:
Fix docs

pamarcos committed:
Enable XRay for ARM

pamarcos committed:
Use more LowCardinality when it makes sense

pamarcos committed:
Add more documentation

pamarcos committed:
Don't use XRay along with sanitizer for ARM since they're not compatible

pamarcos committed:
Fix style check

pamarcos committed:
Minor change

pamarcos committed:
Merge branch 'master' into xray_experiments

pamarcos committed:
Lazily parse the XRay instrumentation map only when trying to set up a new handler

pamarcos committed:
Add stress test

pamarcos committed:
Fix for patching/unpatching even if some other handler is installed

pamarcos committed:
Add log traces to help debugging

pamarcos committed:
Reorder functions

pamarcos committed:
Fix race condition because SharedLockGuard is not what I thought :)

pamarcos committed:
Minor improvement

pamarcos committed:
Rename XRayInstrumentation to just Instrumentation

pamarcos committed:
Improve comment of stress test

pamarcos committed:
Merge branch 'master' into xray-instrumentation

pamarcos committed:
Rename system.instrumentation_profiling_log to system.instrumentation_trace_log

pamarcos committed:
Rename docs after name refactor

pamarcos committed:
More renames

pamarcos committed:
Add support for running a subquery in SYSTEM INSTRUMENT REMOVE statement

pamarcos committed:
Sort new instrumentation section in doc

pamarcos committed:
Rename method to keep it simpler

pamarcos committed:
Migrate to boost::multi_index to use a single container for instrumented points

pamarcos committed:
Fix bug due to race condition in profile handler

pamarcos committed:
Merge branch 'master' into xray-instrumentation

pamarcos committed:
Make parsing of XRay instrumentation single-threaded

Turns out that it's dangerous to use a different LLVMSymbolizer
per thread because it consumes a lot of memory. When optimizing
so that each thread takes a function to symbolize, I realized that
when all cores were symbolizing at the same time the AWS dev machine
ended up unresponsive. There are some workarounds:

1. call LLVMSymbolizer.flush() once in a while to drop the cache.
   Take into account that calling pruneCache() is not enough because
   there are some other internal caches that use a lot of memory.
2. Use a single LLVMSymbolizer and a mutex. This makes the code more
   complex since there are still several threads involved. The worst
   part is that there is no performance gain since the background
   threads have lower priority than the one where the query is executed.

In summary, it's difficult to predict how much memory LLVMSymbolizer
ends up using, but the theory is that each each will use at least the
DWARF information chunk from the binary, which for a debug buidl is big.

In conclusion, I don't think having parallel threads working on it is
worth the extra complexity and the danger to leave the machine without
memory to work with.

pamarcos committed:
Merge branch 'master' into xray-instrumentation

pamarcos committed:
Remove unnecessary headers

pamarcos committed:
Fix messup during merge

pamarcos committed:
Merge branch 'master' into xray-instrumentation

pamarcos committed:
Clarify XRay config message

pamarcos committed:
Add trace to system.instrumentaton_trace_log with the stack

pamarcos committed:
Improve instrumentation logging

pamarcos committed:
Fix typo

serxa commented on src/Interpreters/InstrumentationManager.cpp:
```diff
@@ -0,0 +1,530 @@
+            initialization_status.wait(InitializationStatus::INITIALIZING);
+            LOG_DEBUG(logger, "InstrumentationManager initialized by some other query. Setting up handler");
+        }
+    }
```
There is `Common/callOnce.h`, I believe it should do this

pamarcos committed:
Improve 14x parsing the XRay instrumentation map

From 21s to 1.5s on my dev machine.
Funny how LLVMSymbolizer is so much slower than our own
implementation.

Kudos to Michael Kolupaev for the idea.

serxa commented on src/Interpreters/InstrumentationManager.cpp:
```diff
@@ -0,0 +1,530 @@
+    if (entry_type == XRayEntryType::TAIL)
+        entry_type = XRayEntryType::EXIT;
+
+    for (const auto & [handler_name, handler_function] : handler_name_to_function)
```
This will always trigger all 3 handlers (and potentially more) to lock mutex and search the same data structure. It would be better to invert the logic: find the tracepoint corresponding to the func_id once, then iterate through all attached handlers if any.

Just my imagination: this will also unlock the possibility of registering multiple handlers of the same type. This may be is pointless with LOG/SLEEP/PROFILE, but would be good for a kind of EXEC with parameters PREDICATE or CODE

pamarcos committed:
Simplify logic to lazy initialize once

Feedback from Sergei Trifonov :)

pamarcos committed:
Fix name for StorageSystemInstrumentation

pamarcos committed:
Fix style

pamarcos committed:
Merge branch 'master' into xray-instrumentation

pamarcos committed:
Add function_id, demangled_symbol and demangled_symbol_stripped to system.symbols

pamarcos committed:
Rename unit test

pamarcos committed:
Add complete and demangled symbol name to system.instrumentation

pamarcos committed:
Use substrings to find instrumentation points

Fix setting instrumentation points on templated functions.

This simplifies quite a bit the code and makes it less error prone
to C++ parsing shenanigans.
Since we're now showing the complete symbol name both in
system.instrumentation and in system.symbols, we expect the users
to use a name as close as possible to the complete one.

pamarcos committed:
Prevent reentrancy to avoid instrumenting while instrumenting

pamarcos commented on src/Interpreters/InstrumentationManager.cpp:
```diff
@@ -0,0 +1,499 @@
+    return points;
+}
+
+void InstrumentationManager::dispatchHandler(Int32 func_id, XRayEntryType entry_type)
```
Excellent point. Thanks for finding this out :ok_hand: 
Done at https://github.com/ClickHouse/ClickHouse/pull/89173/commits/009faa385bb33b1bcd45d436bcafa495ac748f69

pamarcos committed:
Add cpu_id to system.trace_log

pamarcos committed:
Fix style check

pamarcos committed:
Move system.instrumentation_trace_log to system.trace_log

pamarcos committed:
Add test for system.symbols

pamarcos committed:
Improve test

pamarcos committed:
Fix build when not using XRay

pamarcos committed:
Include whole stacktrace and improve tests

pamarcos committed:
Always include instrumentation values in system.trace_log to simplify

pamarcos committed:
Use Enum for entry_type and update doc

pamarcos committed:
Merge branch 'master' into xray-instrumentation

pamarcos committed:
Fix style check

pamarcos committed:
Update documentation

pamarcos committed:
Fix docs

pamarcos committed:
Fix build for macOS

pamarcos committed:
Fix profile event in system.trace_log

pamarcos committed:
Merge branch 'master' into xray-instrumentation

pamarcos commented:
I'm generating some profile traces with this script:

```bash
#!/bin/bash

readonly CLICKHOUSE_CLIENT="/home/ubuntu/ClickHouse/ClickHouse/build/programs/clickhouse client"

$CLICKHOUSE_CLIENT -q """
    SYSTEM INSTRUMENT REMOVE ALL;
    TRUNCATE TABLE system.trace_log;
    SYSTEM INSTRUMENT ADD \`sleepForNanoseconds\` PROFILE;
"""

function cleanup()
{
    $CLICKHOUSE_CLIENT -q "SYSTEM INSTRUMENT REMOVE ALL;"
}

trap cleanup EXIT

function query()
{
    for _ in $(seq 1 10); do
        random=$($CLICKHOUSE_CLIENT -q "SELECT (rand() % 1000) / 1000")
        $CLICKHOUSE_CLIENT -q "SELECT sleep($random) FORMAT Null;"
    done
}

for _ in $(seq 1 $(nproc)); do
    query &
done

wait
```

Then converting to Chrome trace format using this chrome_trace.sql file:

```sql
WITH traces AS (
    SELECT * FROM system.trace_log
    WHERE event_date >= today() AND trace_type = 'Instrumentation' AND handler = 'profile'
    ORDER BY event_time, entry_type
)
SELECT
    format(
        '{{"traceEvents": [{}\n]}}',
        arrayStringConcat(
            groupArray(
                format(
                    '\n{{"name": "{}", "cat": "clickhouse", "ph": "{}", "ts": {}, "pid": 1, "tid": {}, "args": {{"query_id": "{}", "cpu_id": {}, "stack": [{}]}}}},',
                    function_name,
                    if(entry_type = 0, 'B', 'E'),
                    timestamp_ns/1000,
                    toString(thread_id),
                    query_id,
                    cpu_id,
                    arrayStringConcat(arrayMap((x, y) -> concat('"', x, ': ', y, '", '), lines, symbols))
                )
            )
        )
    )
FROM traces;
```

```bash
 echo $(clickhouse client --query "$(cat chrome_trace.sql)") > trace.json
```

[trace.zip](https://github.com/user-attachments/files/23482712/trace.zip)

Opening the trace in https://ui.perfetto.dev/

<img width="3014" height="1894" alt="image" src="https://github.com/user-attachments/assets/cd82956b-c75d-4862-aaf7-c8e85fd6321a" />

pamarcos committed:
Another fix of the doc

pamarcos committed:
Fix test

pamarcos committed:
Change parameters to Array(Dynamic)

pamarcos committed:
Update tests

pamarcos committed:
Allow adding more instrumentation points of the same handler
to the same function

pamarcos committed:
Add test for multiple handlers to the same symbol

pamarcos committed:
Add random sleep time option

pamarcos committed:
Fix cpp style check

pamarcos committed:
Separate parameters by whitespace and use integer milliseconds

pamarcos committed:
Fix style

al13n321 committed:
Fix sporadic llvm-project compilation errors

ahmadov commented on docs/en/operations/system-tables/trace_log.md:
```diff
@@ -61,14 +67,69 @@ SELECT * FROM system.trace_log LIMIT 1 \G
+revision:                54504
+trace_type:              Instrumentation
+cpu_id:                  19
+thread_id:               3166432 -- 3.17 million
```
` -- 3.17 million` is redundant

ahmadov commented on docs/en/operations/system-tables/trace_log.md:
````diff
@@ -61,14 +67,69 @@ SELECT * FROM system.trace_log LIMIT 1 \G
+duration_microseconds:   58
 ```
+
+The profiling information can be converted easily to Chrome's Event Trace Format creating the following query in a `chrome_trace.sql` file:
````
maybe?

The profiling data can be extract to a format accepted by Google Chrome Tracing tools with the following SQL query:

pamarcos committed:
Use hashed index for function_id

pamarcos committed:
Make the sleep test faster

pamarcos committed:
Fix to make sure handlers are executed in the same order

The test was working ok by sheer luck because the handlers
added were exactly in alphabetical order.

pamarcos committed:
Fix printing the parameters of the AST for SYSTEM INSTRUMENT

ahmadov commented on src/Interpreters/InstrumentationManager.h:
```diff
@@ -0,0 +1,154 @@
+
+#if USE_XRAY
+
+#include <vector>
```
I don't think we have a preferred include order in the codebase, but what do you think about the following order?
```cpp
<CH_includes>

<external-libraries>

<cpp-libraries>
```

pamarcos replied:
Same, I don't think we ahve a preferred include order apart of alphabetically sorting the groups. I'll change it as per your suggestion

ahmadov commented on src/Interpreters/InterpreterSystemQuery.cpp:
```diff
@@ -933,7 +936,23 @@ BlockIO InterpreterSystemQuery::execute()
+#else
+        case Type::INSTRUMENT_ADD:
+        case Type::INSTRUMENT_REMOVE:
+#endif
```
missing `break` or better throw an exception that support is disabled

pamarcos committed:
Fix profile handler for recursive functions

ahmadov commented on src/Interpreters/TraceLog.cpp:
```diff
@@ -245,6 +264,12 @@ void TraceLogElement::appendToBlock(MutableColumns & columns) const
         columns[i++]->insertDefault();
     }
+
+    columns[i++]->insert(function_id.has_value() ? function_id.value() : Field());
```
```suggestion
    columns[i++]->insert(function_id.value_or(Field()));
```

ahmadov commented on src/Interpreters/TraceLog.cpp:
```diff
@@ -5,15 +5,18 @@
 #include <Common/SymbolIndex.h>
 #include <Common/Dwarf.h>
 #include <IO/WriteBufferFromArena.h>
+#include <xray/xray_interface.h>
```
missing `#if USE_XRAY`?

ahmadov commented on src/Interpreters/InstrumentationManager.cpp:
```diff
@@ -0,0 +1,482 @@
+static constexpr String LOG_HANDLER = "log";
+static constexpr String PROFILE_HANDLER = "profile";
+
+auto logger = getLogger("InstrumentationManager");
```
can we make it a member of use `static`?

pamarcos commented on src/Common/TraceSender.cpp:
```diff
@@ -61,6 +65,11 @@ void TraceSender::send(TraceType trace_type, const StackTrace & stack_trace, Ext
     WriteBufferFromFileDescriptorDiscardOnFailure out(pipe.fds_rw[1], buf_size, buffer);
 
     std::string_view query_id;
+#ifdef OS_LINUX
```
@Ergus asked me to include it because he uses some tool to visualize profiling data that includes it. I would have also assumed therad_id was enough, but it's no big deal to include also the cpu_id and might be useful to map which thread_ids run in which cpu_ids. For instance, to analyze affinity stuff.

pamarcos committed:
Update src/Interpreters/TraceLog.cpp

Co-authored-by: Elmi Ahmadov <elmi.1295@gmail.com>

pamarcos committed:
Update src/Interpreters/InterpreterSystemQuery.cpp

Co-authored-by: Elmi Ahmadov <elmi.1295@gmail.com>

ahmadov reviewed (commented):
I am doing the code review partially, but overall it looks good! I tested it and liked it very much! Good job @pamarcos!

pamarcos committed:
Revert "Update src/Interpreters/TraceLog.cpp"

This reverts commit a7b582cf56467223b822441f923b54861b845b2b.

pamarcos committed:
Sort InstrumentedPointContainer by FunctionId, Id

pamarcos committed:
Improvements from PR's feedback

Ergus commented on src/Interpreters/TraceCollector.cpp:
```diff
@@ -187,6 +191,7 @@ void TraceCollector::run()
                     .event_time_microseconds = time_in_microseconds,
                     .timestamp_ns = timestamp_ns,
                     .trace_type = trace_type,
+                    .cpu_id = cpu_id,
```
;)

Ergus commented on src/Interpreters/TraceLog.cpp:
```diff
@@ -245,6 +264,12 @@ void TraceLogElement::appendToBlock(MutableColumns & columns) const
+    columns[i++]->insert(function_name.has_value() ? function_name.value() : Field());
+    columns[i++]->insert(handler.has_value() ? handler.value() : Field());
+    columns[i++]->insert(entry_type.has_value() ? entry_type.value() : Field());
+    columns[i++]->insert(duration_microseconds.has_value() ? duration_microseconds.value() : Field());
```
better use `value_or`

Ergus reviewed (commented):
We have already discussed some details to consider, but in general this seems like a lot of nice work and the traces look nice ;)

pamarcos committed:
Move get_cpuid() to CPUID.h

pamarcos commented on src/Interpreters/InstrumentationManager.cpp:
```diff
@@ -0,0 +1,484 @@
+
+    if (entry_type == XRayEntryType::ENTRY)
+    {
+        auto element = createTraceLogElement(instrumented_point, entry_type, std::chrono::system_clock::now());
```
To have a more precise measurement. There's a log and an if branch in between.

pamarcos committed:
Simplify logic removing unnecessary std::optional

Also use Instrumentation::EntryType instead of XRay's enum
to encapsulate it ourselves.

pamarcos committed:
Simplify logic following PR's feedback

pamarcos committed:
Merge branch 'master' into xray-instrumentation

pamarcos commented on src/Interpreters/InterpreterSystemQuery.cpp:
```diff
@@ -1732,6 +1751,83 @@ void InterpreterSystemQuery::loadOrUnloadPrimaryKeysImpl(bool load)
+        {
+            if (!query.instrumentation_subquery.empty())
+            {
+                auto [_, block_io] = executeQuery(query.instrumentation_subquery, getContext(), QueryFlags{ .internal = true });
```
I'm not sure I follow you. Do you mind elaborating what you mean?

ahmadov replied:
sure, instrumentation_subquery is executed and later it's assumed that the result type is `UInt`: https://github.com/ClickHouse/ClickHouse/blob/f5ceaae8704ada91c55e73a0f3fbcf28bfe4b85c/src/Interpreters/InterpreterSystemQuery.cpp#L1788.

I don't if it's possible to check the result type before instrumentation_subquery execution. .e.g

```cpp
if (getReturnType(query.instrumentation_subquery) != UInt)
   throw /* unexpected type */;
auto [_, block_io] = executeQuery(...)
```

something like that.

pamarcos committed:
Relax a little bit the stack trace requirements for the test

From https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=89173&sha=b4c91173018a17d26210f2c116f5c142784cb9cb&name_0=PR
we can see that there are times where there's no
`QueryMetricLog::startQuery` even on entry, for some reason.

SELECT
    event_time,
    handler,
    entry_type,
    function_name,
    symbols
FROM file('trace_log.tsv.zst')
WHERE (trace_type = 'Instrumentation') AND (query_id = 'test_7ja0j7ii_profile') AND (function_name LIKE '%QueryMetricLog::startQuery%')

Query id: 45c33195-f2f9-499b-b52c-03c2d7ba4e94

Row 1:
──────
event_time:    2025-11-14 02:03:08
handler:       profile
entry_type:    Entry
function_name: DB::QueryMetricLog::startQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::chrono::time_point<std::__1::chrono::system_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000l>>>, unsigned long)
symbols:       ['StackTrace::StackTrace()','DB::InstrumentationManager::createTraceLogElement(DB::InstrumentationManager::InstrumentedPointInfo const&, XRayEntryType, std::__1::chrono::time_point<std::__1::chrono::system_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000l>>>) const','DB::InstrumentationManager::profile(XRayEntryType, DB::InstrumentationManager::InstrumentedPointInfo const&)','DB::InstrumentationManager::dispatchHandlerImpl(int, XRayEntryType)','DB::InstrumentationManager::dispatchHandler(int, XRayEntryType)']

Row 2:
──────
event_time:    2025-11-14 02:03:08
handler:       profile
entry_type:    Exit
function_name: DB::QueryMetricLog::startQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::chrono::time_point<std::__1::chrono::system_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000l>>>, unsigned long)
symbols:       ['StackTrace::StackTrace()','DB::InstrumentationManager::createTraceLogElement(DB::InstrumentationManager::InstrumentedPointInfo const&, XRayEntryType, std::__1::chrono::time_point<std::__1::chrono::system_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000l>>>) const','DB::InstrumentationManager::profile(XRayEntryType, DB::InstrumentationManager::InstrumentedPointInfo const&)','DB::InstrumentationManager::dispatchHandlerImpl(int, XRayEntryType)','DB::InstrumentationManager::dispatchHandler(int, XRayEntryType)']

pamarcos committed:
Add test for profiling a recursive function

pamarcos committed:
Improve resolution when profiling to nanoseconds

pamarcos committed:
Avoid using std::optional unless it's completely needed

pamarcos commented on src/Interpreters/TraceLog.h:
```diff
@@ -45,6 +46,14 @@ struct TraceLogElement
+    /// Instrumentation values
+    UInt64 instrumented_point_id;
+    std::optional<Int32> function_id{};
+    std::optional<String> function_name{};
```
Done in https://github.com/ClickHouse/ClickHouse/pull/89173/commits/5fb2b5fece4567f2350058f5069299ccc21638aa

pamarcos commented:
I've checked the function alignment after @azat comment about the fact that [we're aligning at 64 bytes](https://github.com/ClickHouse/ClickHouse/blob/65813e712e82839cb4fe6600687da1ccbb784c91/CMakeLists.txt#L285-L286) the start of every function to make the performance more stable.

Having this file, compiled with `clang++ -fxray-instrument -fxray-instruction-threshold=1 -falign-functions=64 -ggdb3 test_xray.cpp -o test_xray`:

```cpp
#include <cstdint>
#include <cstdio>
#include <xray/xray_interface.h>

[[clang::xray_never_instrument]] void my_handler(int32_t fid, XRayEntryType type)
{
    printf("arg0: FuncID: %d, Type: %d\n", fid, type);
}

void foo(int * value, int other_value)
{
    printf("Processing: %d (%lu) and %d\n", *value, (uint64_t)value, other_value);
}

int main()
{
    __xray_init();

    // Patch all instrumented functions
    __xray_patch();

    // Set the handler
    __xray_set_handler(my_handler);

    // Call instrumented functions
    int value = 10;
    foo(&value, 20);

    // Cleanup
    __xray_remove_handler();
    __xray_unpatch();

    return 0;
}
```

```bash
objdump -D test_xray | c++filt | grep -A 5 foo
0000000000023400 <foo(int*, int)>:
   23400:       eb 09                   jmp    2340b <foo(int*, int)+0xb>
   23402:       66 0f 1f 84 00 00 02    nopw   0x200(%rax,%rax,1)
   23409:       00 00
   2340b:       55                      push   %rbp
   2340c:       48 89 e5                mov    %rsp,%rbp
   2340f:       48 83 ec 10             sub    $0x10,%rsp
```

We can see that as expected, the instrumentation is added first ([and it's actually aligned to 16 bytes](https://github.com/ClickHouse/llvm-project/blob/28fde6ac54a46f8ec9b27dbca903f817cfa427b5/compiler-rt/lib/xray/xray_trampoline_x86_64.S#L18-L23)) and then the alignment to 64 bytes is applied. So, every function is now starting with an alignment of 64+11=75 bytes for x86_64. I believe this alignment difference might be the root cause of the performance degradation of `sum_map`, because @serxa also commented it's very sensitive to any change.

I want to test the same for aarch64, but I imagine the same would happen. But I'd like to verify how many bytes are added before and after.

pamarcos committed:
Merge branch 'master' into xray-instrumentation

pamarcos committed:
Enable XRay by default except for sanitized builds

pamarcos committed:
Fix build after merging master

pamarcos committed:
Include XRay instrumentation only if LLVM is used

pamarcos committed:
Use tag with hyphen for consistency

pamarcos committed:
Revert "Include XRay instrumentation only if LLVM is used"

This reverts commit 0ebc775cdf6cdcbb6cb811c502e6455681148764.

pamarcos committed:
Fix XRay cmake config after enabling it by default

pamarcos committed:
Remove unnecessary offset applied to traces

pamarcos committed:
Add error handling for wrong type result in subquery execution

pamarcos committed:
Add trailing comma only if needed

pamarcos committed:
Add TSA annotation

pamarcos committed:
Clarify variable name

pamarcos committed:
Simplify logic

pamarcos committed:
Address comments from Copilit's PR feedback

pamarcos committed:
Fix build for builds without XRay

pamarcos committed:
Disable XRay for musl flavor

pamarcos committed:
Merge branch 'master' into xray-instrumentation

pamarcos committed:
Merge branch 'master' into xray-instrumentation

pamarcos committed:
Disable XRay modes except the core library for function entry and exit

This way we make sure that by default only function entry and exit
slopes are emitted. Also, we avoid linking unnecessarily xray-basic
and xray-fdr.

ahmadov reviewed (approved):
overall lgtm.

pamarcos committed:
Merge branch 'master' into xray-instrumentation

pamarcos committed:
Throw exception in case of no support for XRay

pamarcos committed:
Build AMD compat with AMD instead of cross-compiling due to XRay lib missing

pamarcos committed:
fixup! Throw exception in case of no support for XRay

pamarcos committed:
Fix to run AMD compat on AMD

I tried to fix it in a previous commit but changed the wrong
build flavor :).

pamarcos committed:
Merge branch 'master' into xray-instrumentation

pamarcos committed:
Run AMD_COMPAT with ARM

pamarcos committed:
fixup! Run AMD_COMPAT with ARM

pamarcos commented:
I'm merging this after syncing with master because cross compilation builds affected slightly how XRay flavor needed to be built.
I know there's a performance degradation in `sum_map` , but I decided to merge anyway to catch what other performance issues we may find.

Algunenano commented:
Moved the discussion to an issue: https://github.com/ClickHouse/ClickHouse/issues/91475

azat commented on src/Parsers/ParserSystemQuery.cpp:
```diff
@@ -748,6 +751,92 @@ bool ParserSystemQuery::parseImpl(IParser::Pos & pos, ASTPtr & node, Expected &
+                break;
+            }
+
+            res->instrumentation_point_id = temporary_identifier->as<ASTLiteral>()->value.safeGet<UInt64>();
```
It will be nice to support removing by identifier/string, but my point here was that the exception is unexpected, it should be some kind of parsing error

pamarcos commented on src/Parsers/ParserSystemQuery.cpp:
```diff
@@ -748,6 +751,92 @@ bool ParserSystemQuery::parseImpl(IParser::Pos & pos, ASTPtr & node, Expected &
+        case Type::INSTRUMENT_ADD:
+        {
+            ASTPtr temporary_identifier;
+            if (ParserIdentifier{}.parse(pos, temporary_identifier, expected))
```
Done in https://github.com/ClickHouse/ClickHouse/pull/93345
